File: CarePackage/Software/JetBrainsToolbox.cs
namespace CarePackage.Software;

public class JetBrainsToolbox : BaseSoftware
{
    public override string           Key            { get; set; } = "jetbrains-toolbox";
    public override string           Name           { get; set; } = "JetBrains Toolbox";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Development;
    public override string           DownloadName   { get; set; } = "jetbrains-toolbox.exe";
    public override bool             IsArchive      { get; set; } = false;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = false;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.jetbrains_toolbox;
    public override string           Homepage       { get; set; } = "https://jetbrains.com/toolbox-app";

    private readonly HttpClient _http;
    
    public JetBrainsToolbox(IServiceProvider services)
    {
        _http = services.GetKeyedService<HttpClient>("MimicBrowser")!;
    }

    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var res = await _http.GetAsync("https://data.services.jetbrains.com/products/releases?code=TBA&latest=true&type=release", ct);

        res.EnsureSuccessStatusCode();

        var downloadUrlPattern = new Regex(@"https://download\.jetbrains\.com/toolbox/jetbrains-toolbox-\d+\.\d+\.\d+\.\d+\.exe\b");
        var html               = await res.Content.ReadAsStringAsync(ct);
        var match              = downloadUrlPattern.Match(html);
        
        DownloadUrlResolveException.ThrowUnless(match.Success);

        return match.Groups[0].Value;
    }
}

File: packages/app/src/services/software/definitions/jetbrainsToolbox.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'jetbrains-toolbox';
	public name = 'JetBrains Toolbox';
	public category = [SoftwareCategory.Development];
	public downloadName = 'jetbrains-toolbox.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'jetbrains-toolbox.png';
	public homepage = 'https://jetbrains.com/toolbox-app';

	public async resolveDownloadUrl() {
		const res = await fetch('https://data.services.jetbrains.com/products/releases?code=TBA&latest=true&type=release');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/download\.jetbrains\.com\/toolbox\/jetbrains-toolbox-\d+\.\d+\.\d+\.\d+\.exe\b/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}
